projects/packages/assets/src/class-initial-state.php

```diff
@@ -0,0 +1,160 @@
+<?php
+/**
+ * Jetpack Initial state.
+ *
+ * @package  automattic/jetpack-assets
+ */
+
+namespace Automattic\Jetpack;
+
+use Jetpack_Options;
+
+/**
+ * Class Initial State
+ */
+class Initial_State {
+
+	const SCRIPT_HANDLE = 'jetpack-initial-state';
+
+	/**
+	 * The singleton instance of this class.
+	 *
+	 * @var Initial_State
+	 */
+	protected static $instance;
+
+	/**
+	 * Constructor.
+	 *
+	 * Static-only class, so nothing here.
+	 */
+	private function __construct() {}
+
+	/**
+	 * Get the singleton instance of the class.
+	 *
+	 * @return Initial_State
+	 */
+	public static function instance() {
+		if ( ! isset( self::$instance ) ) {
+			self::$instance = new Initial_State();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Configure.
+	 */
+	public function configure() {
+		// Ensure that assets are registered on wp load,
+		// so that when dependent scripts are enqueued, the scripts here are already registered.
+		add_action( 'wp_loaded', array( $this, 'register_assets' ) );
+
+		// We want to render the initial state as late as possible,
+		// so that it can be filtered when plugins want to decide whether they need to enqueue their scripts or not.
+		add_action( 'admin_enqueue_scripts', array( $this, 'render_initial_state' ), 999999 );
+	}
+
+	/**
+	 * Register assets.
+	 */
+	public function register_assets() {
+
+		if ( ! wp_script_is( self::SCRIPT_HANDLE, 'registered' ) ) {
+
+			Assets::register_script(
+				self::SCRIPT_HANDLE,
+				'../build/jetpack-initial-state.js',
+				__FILE__,
+				array(
+					'textdomain' => 'jetpack-assets',
+				)
+			);
+		}
+	}
+
+	/**
+	 * Render the initial state using an inline script.
+	 *
+	 * @return void
+	 */
+	public function render_initial_state() {
+
+		// If the initial state has already been added, don't add it again.
+		// This can happen if this methd is called explicitly.
+		if ( wp_scripts()->get_data( self::SCRIPT_HANDLE, 'Jetpack::InitialState::added' ) ) {
+			return;
+		}
+
+		$initial_state = wp_json_encode(
+			self::get_initial_state(),
+			JSON_UNESCAPED_SLASHES | JSON_HEX_TAG | JSON_HEX_AMP | JSON_UNESCAPED_UNICODE
+		);
+
+		wp_add_inline_script(
+			self::SCRIPT_HANDLE,
+			sprintf( 'var JETPACK_INITIAL_STATE = %s;', $initial_state ),
```

Comment:
I wonder what the merits are of setting this global versus setting a property on the `window.JetpackInitialState` object.

Reply:
I am not sure which is better:

One benefit of `window.JetpackInitialState` is that it can be check in `if()` condition and can also be `deleted`.

Reply:
Since we're already using `window.JetpackInitialState` in both types and utils, can we use it here as well? I don't think there's a downside to it, is there?